### PR TITLE
Detach handshake from connect to node

### PR DESCRIPTION
**Pull request by s1monw**
Today we connect and publish the nodes connection before we execute a
handshake with the node we connect to. In the case of connecting to a node
that won't pass the handshake this connection is already `published` and other
code paths can use it. This commit detaches the connection and the publish of the
connection such that `TransportService` can do a handshake before actually connect
and publish the connection.

### CODE REVIEW

**Comment by s1monw** [comment]
@elasticmachine test this please

**Comment by s1monw** [comment]
@elasticmachine test this please

**Review by jasontedor — COMMENTED** [review]
I like the change @s1monw. I left some comments.

File: core/src/main/java/org/elasticsearch/transport/TransportService.java [hunk 1]
```diff
@@ -330,36 +330,35 @@ public DiscoveryNode connectToNodeAndHandshake(
     public DiscoveryNode connectToNodeAndHandshake(
             final DiscoveryNode node,
             final long handshakeTimeout,
-            final boolean checkClusterName) {
+            final boolean checkClusterName) throws IOException {
```
Comment (s1monw) [inline core/src/main/java/org/elasticsearch/transport/TransportService.java]: I can remove it here but I think we will likely add it back in the future when we try to connect to foreign clusters. Lets trash it it's as simple to add back as it is to remove

**Comment by s1monw** [comment]
@jasontedor I pushed a new commit

**Comment by s1monw** [comment]
@elasticmachine test this please

**Review by jasontedor — APPROVED** [review]
LGTM.

**Comment by s1monw** [comment]
@elasticmachine test this please

**Comment by s1monw** [comment]
@elasticmachine test this please

**Comment by s1monw** [comment]
@elasticmachine test this please

**Pull request merged**